armadillo/src/main/java/at/favre/lib/armadillo/OnSecurePreferenceChangeListener.java

```diff
@@ -0,0 +1,34 @@
+package at.favre.lib.armadillo;
+
+import android.content.SharedPreferences;
+import android.support.annotation.NonNull;
+
+/**
+ * Allow to register preference change listeners that receive
+ * @since 03.07.2019
+ */
+public interface OnSecurePreferenceChangeListener {
+
+    /**
+     * Variation of the regular {@link android.content.SharedPreferences.OnSharedPreferenceChangeListener#onSharedPreferenceChanged(SharedPreferences, String)} that
+     * provides a {@link DerivedKeyComparison} to allow client side react only to a specific changed key
+     * @param comparison utility to check equivalence between plain keys and derived keys
+     * @param sharedPreferences the shared preferences instance
+     * @param derivedKeyChanged the value of the changed key in it's derived form (as it's actually stored)
+     */
+    void onSecurePreferenceChanged(@NonNull DerivedKeyComparison comparison, @NonNull SharedPreferences sharedPreferences, @NonNull String derivedKeyChanged);
+
+    /**
+     * Allows client side {@link OnSecurePreferenceChangeListener} to check if the changed key is the key of interest.
+     */
+    interface DerivedKeyComparison {
```

Comment:
Instead of providing the caller with a generic `doesMyKeyMatchWithTheDerivedKey()` it may be easier to understand for the callback to provide a checker which has the derived key backed in. Therefore the onSecurePreferenceChanged()` would not require to pass the `String derivedKeyChanged` argument and the caller would just check with e.g. `comparison.matchesDerivedKey(String rawContentKey)`

Reply:
Yes! Here's the funny story...

I begun doing it that way indeed, but then I though that in case of positive match I could use the `sharedPreferences` and `key` parameters, to get the changed value.

I then realised that this didn't work because the `sharedPreferences` object was the original one, and requesting a preference with the `key` parameter would provide the _encrypted_ value.

And then I probably left that as it was. I think that it might be nice to have both of the proposals. So let's see what do you think about having the method signature 

`void onSecurePreferenceChanged(@NonNull SharedPreferences sharedPreferences, @NonNull DerivedKeyComparison comparison)`

Where the **instance** of sharedPreferences would be the `SecureSharedPreferences` instead of the underlying `SharedPreferencesImpl` so that client code could then get the changed value.

I didn't do that at the time, worried about security or something similar. If we're not interesting on allowing such behaviour, then I may suggest to change the signature of the method to just receive the `DerivedKeyComparison` parameter as right now is the only usable parameter

Reply:
I like the improved version. What we should change is that the `SharedPreferences sharedPreferences` should be the Armadillo instance not the inner wrapped version. A) a lib user has no business changing the raw encrypted data B) the raw encrypted shared pref is practically useless for the user, with the Armadillo wrapper she would be able to retrieve the enecrypted data with the content-keys she knows.